### DISCUSSION TITLE
Returning null for an entity without extensions error property

### DISCUSSION THREAD

**Discussion by cameronmakarchuk** [hayes/pothos#1171] [Q&A] [answered]
Hi,

I'm using entities in my federated graph and I'm trying to return `null` for the entity ('address' in the example), while keeping the entities fields (i.e. 'city', 'state', 'country') non-nullable, like this:

```
data: {
  user: {
   address: null
  }
}
```

When I query the subgraph that defines the entity directly (Address subgraph in this example), this is no issue and returns `null` without trying to resolve the individual fields. But if I do the same query through a different subgraph (i.e. User subgraph) using externalRef and the Address entity, I get `null` but also the `extensions` property with the error message of 'Cannot return null for non-nullable field Address.city".

```
data: {
  user: {
   address: null
  },
"extensions": {
    "valueCompletion": [
      {
        "message": "Cannot return null for non-nullable field Address.city.",
        "path": [
          "user",
          "address",
          "city"
        ],
        "extensions": {
          "code": "INVALID_GRAPHQL"
        }
      }
    ]
  }
}
```

It's like when going through the externalRef it still tries to resolve the entities fields even though in the referenceResolver I'm returning `null` early if no record is found, and the `address` field is nullable.

I thought it was because I didn't have `resolvable: false` set in the key directive of the externalRef but that didn't make any difference.

Any suggestions on how to do this without also getting the `extensions` property?

**Comment by hayes** [comment]
I can't tell for sure can you share more details of your setup?

For the query that fails, what does.the resolver for the address field return, and what does the the resolveEntity in the graph that implements Address get called with, and what does it return?

**Comment by cameronmakarchuk** [comment]
Yeah for sure, let me see if I can clarify a bit better.

So for this example using User and Address subgraphs, it would look like this:

Address subgraph:
```
const Address= Builder.objectRef<AddressType>('Address').implement({
	fields: (t) => ({
		city: t.exposeString('city'),
		state: t.exposeString('state'),
		country: t.exposeString('country')),
        }),
})

Builder.asEntity(CardTypes.CardDetails, {
	key: Builder.selection<{ id: string }>('id'),
	resolveReference: async ({ id }) => {

		const databaseRecord = <.... get address record from database>

		if (!databaseRecord) {
			return null;
		}

		return {
			data: databaseRecord,
		};
	},
});
```

User subgraph:
```
const Address= Builder.externalRef(
	'Address',
	Builder.keyDirective(Builder.selection<{id: string }>('id'), false),
).implement({
	fields: (t) => ({
		id: t.exposeString('id'),
	}),
});

const User = Builder.prismaObject('User', {
  fields: (t) => ({
        id: exposeString('id'),
        address: t.field({
			type: Address,
			nullable: true,
			resolve: (parent) => {
				const { id } = parent;

				return {
					id,
				};
			},
		}),
       })
})
```

If I run my `user` query like:

```
query User {
   user {
      id,
      address {
          city
      }
  }
}
```

And there is no `databaseRecord` found on the `Address` entity referenceResolver it should return  `null` for `address`, which it does, but I also get an extensions error object like this:

```
data: {
  user: {
    id: 12345
    address: null
  },
"extensions": {
    "valueCompletion": [
      {
        "message": "Cannot return null for non-nullable field Address.city.",
        "path": [
          "user",
          "address",
          "city"
        ],
        "extensions": {
          "code": "INVALID_GRAPHQL"
        }
      }
    ]
  }
}
```

I just can't seem to figure out how to get that `extensions` field removed from the response.

**Reply by hayes** [reply, marked as answer]
I think this is just how Federation works, your User graph is returning an Address not null, so at that point the router thinks the User has an address, and failing to load it is an Error.

The way to avoid this is to invert the relation and move the User.address field to the Address sub-graph

**Reply by cameronmakarchuk** [reply]
Ok thanks for the follow up, that's kind of what I thought. I'll play around with inverting the relation and see if I can make that work. Thanks!

**Comment by cameronmakarchuk** [comment]
In this example, if I were to do the same query to the Address subgraph directly:

```
query Address {
   address {
      id,
      city,
  }
}
```

And the `databaseRecord` is not found, it would return `null` without the `extensions` property as expected like this:

```
data: {
  address: null,
}
```

That `extensions` property only comes back on the response when going through the entity.
